File: Python/Searching and Sorting/01_first_and_last_occurence_of_x.py
"""
Idea:
Using Modified binary search, finding 1st & last occurrence of a number
-By finding element to left/right even after finding it in 1st try, till corner element is found
"""


class Solution:
    def find_first_occurrence(self, arr, x):
        left, right = 0, len(arr) - 1
        result = -1

        while left <= right:
            mid = left + (right - left) // 2

            if arr[mid] == x:
                result = mid
                right = mid - 1  # Look in the left subarray for the first occurrence
            elif arr[mid] < x:
                left = mid + 1
            else:
                right = mid - 1

        return result

    def find_last_occurrence(self, arr, x):
        left, right = 0, len(arr) - 1
        result = -1

        while left <= right:
            mid = left + (right - left) // 2

            if arr[mid] == x:
                result = mid
                left = mid + 1  # Look in the right subarray for the last occurrence
            elif arr[mid] < x:
                left = mid + 1
            else:
                right = mid - 1

        return result

    def find(self, arr, n, x):
        return self.find_first_occurrence(arr, x), self.find_last_occurrence(arr, x)


t = int(input())
for _ in range(0, t):
    l = list(map(int, input().split()))
    n = l[0]
    x = l[1]
    arr = list(map(int, input().split()))
    ob = Solution()
    ans = ob.find(arr, n, x)
    print(*ans)

File: C++/Searching and Sorting/01_first_and_last_occurence_of_x.cpp
/*
Idea:
Using Modified binary search, finding 1st & last occurence of a number
*/


#include<bits/stdc++.h>
using namespace std;


class Solution
{
public:
    int findLeft(int arr[], int n, int x)
    {
        int low = 0, high = n - 1, result = -1;

        while (low <= high)
        {
            int mid = low + (high - low) / 2;
            if (arr[mid] == x)
            {
                result = mid;
                high = mid - 1; //  Look in the left subarray for the first occurrence
            }
            else if (arr[mid] < x)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return result;
    }

    int findRight(int arr[], int n, int x)
    {
        int low = 0, high = n - 1, result = -1;
        while (low <= high)
        {
            int mid = low + (high - low) / 2;
            if (arr[mid] == x)
            {
                result = mid;
                low = mid + 1; //  Look in the right subarray for the last occurrence
            }
            else if (arr[mid] < x)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return result;
    }
    vector<int> find(int arr[], int n, int x)
    {
        vector<int>ans;
        ans.push_back(findLeft(arr, n, x));
        ans.push_back(findRight(arr, n, x));
        return ans;
    }
};



int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, x;
        cin >> n >> x;
        int arr[n], i;
        for (i = 0;i < n;i++)
            cin >> arr[i];
        vector<int> ans;
        Solution ob;
        ans = ob.find(arr, n, x);
        cout << ans[0] << " " << ans[1] << endl;
    }
    return 0;
}


